### ISSUE TITLE
Forbid name shadowing 

### ISSUE THREAD

**Issue by sobolevn** [wemake-services/wemake-python-styleguide#426]
# Rule request

## Thesis

We can shadow names when using nested functions.
Like in this example:

```python
def test():
    test = 1

    def other(test):
        print(test)

    other(2)

test()
```

Can you tell what will be printed from the first sight?

## Reasoning

Name shadowing can be very tricky. And it will cause different errors in your code.
We need to prevent that.

**Comment by kxepal** [comment]
That's easy - 2. You just start from local context and getting through outer ones until hit globals. This example is too syntetic, but consider this one:
```
def any(thing):
  thing = some(thing)
  def no(thing):
    pass
  return no(thing)
```
We want to explicitly pass `thing` to inner no function, but it still named the same because it means the same outer one.

**Comment by sobolevn** [comment]
Yes, it is. There's no need to:

1. Use nested functions generally (we only allow factories and decorators)
2. Use explicit parameters instead of closures
3. Naming it the same way as original parameter, since it is already another thing, because it now acts in different context

**Comment by kxepal** [comment]
Well, if you forbid point 1, I don't see any cases when this issue may exists. However, there would be need to detect decorator-functions. But their pattern +/- quite oblivious.

**Comment by sobolevn** [comment]
@kxepal we already forbid most nested functions except `decorator` and `factory` functions. See https://wemake-python-styleguide.readthedocs.io/en/latest/pages/violations/best_practices.html#wemake_python_styleguide.violations.best_practices.NestedFunctionViolation

**Comment by kxepal** [comment]
Awesome! Than that this issue shouldn't be an issue at all. What's the case?

**Comment by sobolevn** [comment]
For `decorator` and `factory` functions that can be nested 🙂

**Comment by kxepal** [comment]
But only for one (simple decorators) or two levels (paratetrized decorators). Nothing deeper and there closure is isued to by pass arguments, so no shadowing happens.

**Comment by kxepal** [comment]
_Do you have rule to forbid `# noqa` commetaries? That should be helpful._

**Comment by sobolevn** [comment]
Yes, I have a plan to forbid too many `noqa` comments.

**Comment by sobolevn** [comment]
Related #399 

I will close this issue in order to work on #399 with bigger scope.
